cmake/CheckTrezor.cmake

```diff
@@ -102,13 +113,14 @@ if(Protobuf_FOUND AND USE_DEVICE_TREZOR AND TREZOR_PYTHON)
         "${CMAKE_BINARY_DIR}/test-protobuf.pb.cc"
         "${CMAKE_CURRENT_LIST_DIR}/test-protobuf.cpp"
         CMAKE_FLAGS
+        CMAKE_EXE_LINKER_FLAGS ${CMAKE_TRY_COMPILE_LINKER_FLAGS}
         "-DINCLUDE_DIRECTORIES=${Protobuf_INCLUDE_DIR};${CMAKE_BINARY_DIR}"
-        "-DCMAKE_CXX_STANDARD=11"
-        LINK_LIBRARIES ${Protobuf_LIBRARY}
+        "-DCMAKE_CXX_STANDARD=14"
+        LINK_LIBRARIES ${Protobuf_LIBRARY} ${CMAKE_TRY_COMPILE_LINK_LIBRARIES}
         OUTPUT_VARIABLE OUTPUT
     )
     if(NOT Protobuf_COMPILE_TEST_PASSED)
-        message(STATUS "Protobuf Compilation test failed: ${OUTPUT}.")
+        message(FATAL_ERROR "Trezor: Protobuf Compilation test failed: ${OUTPUT}.")
```

Comment:
This breaks development builds on Arch Linux, Brew, MSYS2, which all ship Abseil built with C++17.

https://github.com/abseil/abseil-cpp/issues/696#issuecomment-636949929

Reply:
Do you pls have any CI link for this? All tests are passing here. Trezor support is mandatory now, unless it is explicitly turned off by compile switch. 

Also, when compile test fails, is the whole build passing with Trezor?

Reply:
Rerun CI. MSYS2 updated to Protobuf 23.4 (from 21.23) 5 days ago.

Reply:
Hmm, what do you propose? Can we pin protobuf to v 21? 
@selsta pls any update on c++17? Thanks!

Reply:
I opened a PR for it but it's unclear if others agree with it: https://github.com/monero-project/monero/pull/8922

Reply:
>     Disable Trezor support by default if Protobuf > 21.

This would result in official builds without Trezor support. Weird solution, rather pin protobuf to v21 or install from sources

Reply:
AFAIK only win build is broken here. And according to the https://www.msys2.org/docs/package-management/ and https://github.com/msys2/setup-msys2 we can install older protobuf version explicitly. This handles CI case.
 (=official releases).

If user wants to do it on their own, they can a) disable Trezor support if they dont need it b) wait for c++17 support c) follow CI recipe and install compatible protobuf.

Reply:
Official CLI builds are built with `depends` and it [doesn't use](https://github.com/monero-project/monero/blob/master/contrib/depends/packages/native_protobuf.mk#L2) Protobuf > 21. Monero GUI builds are pinned at an older version too.

FYI, the monero-gui Arch Linux [package](https://gitlab.archlinux.org/archlinux/packaging/packages/monero-gui/-/blob/main/PKGBUILD) (which is built against Protobuf 23.4) does not have Trezor support because of the broken test in CheckTrezor. So this should be fixed regardless.

Reply:
MSYS2 does not support downgrades. It would be easier to configure CI to 
```
sed -e 's|CMAKE_CXX_STANDARD 14|CMAKE_CXX_STANDARD 17|' -i CMakeLists.txt
``` 
and fix the test.

Reply:
|                             | Current state           | With this PR                                                         | 
|-----------------------------|-------------------------|----------------------------------------------------------------------|
| Official builds  (Pb <= 21) | ✅ Trezor support (not enforced) | ✅ Trezor support (enforced by default)                                             |
| Dev builds (Pb <= 21)       | ✅ Trezor support (not enforced) | ✅ Trezor support (enforced by default)                                               |
| Dev builds (Pb > 21) [0]       | ❌ Trezor support, [1]| ❌ **Builds fail**, [2]                   |
| Monero packages (Pb > 21)   | ❌ Trezor support, [1] | ❌ **Builds fail**, [1] |

[0] Currently: Arch Linux (and derivatives), MSYS2, and Homebrew (unless Pb is downgraded).

[1] To enable Trezor support user/maintainers need to set `CMAKE_CXX_STANDARD` to `17` and remove the test in `CheckTrezor.cmake`. (https://paste.debian.net/plainh/6361dd42)

[2] To disable Trezor support users need to manually pass `-DUSE_DEVICE_TREZOR=OFF`. 

A less drastic measure to require Trezor support for official **CLI** builds would be to check if we're building a release build with `if(DEPENDS)` and force it only then. Unfortunately, this doesn't work for the GUI as it doesn't use `depends` for all builds. Alternatively, you can check if `CMAKE_BUILD_TYPE` is `Release`, but this will catch some users building with `make release`, so this isn't ideal.

My point is that I would rather see graceful failure for Protobuf > 21, because it will otherwise result in an influx of support queries.

Reply:
hmm @tobtoht I see, but what do you suggest? :) I could add a switch to disable compilation test. But is there a way to detect protobuf is > 21?

Reply:
https://cmake.org/cmake/help/latest/module/FindProtobuf.html has a Protobuf_VERSION

Reply:
@selsta @tobtoht I've added a new switch USE_DEVICE_TREZOR_PROTOBUF_TEST, enabled by default. It checks if protobuf > 21, then it fails gracefully. Otherwise continues with the compilation test. When set to OFF, no Protobuf version check and compile test is done. Is this ok?